geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegion.java

```diff
@@ -1221,24 +1221,36 @@ public void addGatewaySenderId(String gatewaySenderId) {
   }
 
   public void updatePRConfigWithNewSetOfGatewaySenders(Set<String> gatewaySendersToAdd) {
-    PartitionRegionHelper.assignBucketsToPartitions(this);
-    PartitionRegionConfig prConfig = this.prRoot.get(getRegionIdentifier());
-    prConfig.setGatewaySenderIds(gatewaySendersToAdd);
-    updatePRConfig(prConfig, false);
+    final RegionLock rl = getRegionLock();
+    rl.lock();
+    try {
+      PartitionRegionHelper.assignBucketsToPartitions(this);
```

Comment:
I would vote for moving the call to assignBucketsToPartitions outside your no lock code.
I think you just want to hold the lock while reading and modifying prConfig.
assignBucketsToPartitions gets its own lock (the recovery lock) and I don't see a logical reason why you would want to do it while locked.
Consider refactoring the lock, unlock, the get of the preconfig and the updatePRConfig calls into a single method that takes an interface that modifies a pr config: "interface PartitionRegionConfigModifier {void modify(PartitionRegionConfig)". You could then call that method in both of these with just a lambda that does the actual change of the prconfig.